lib/reports/matched_content_changes_report.rb

```diff
@@ -0,0 +1,43 @@
+class Reports::MatchedContentChangesReport
+  OUTPUT_ATTRIBUTES = {
+    created_at: :content_change,
+    base_path: :content_change,
+    change_note: :content_change,
+    document_type: :content_change,
+    publishing_app: :content_change,
+    priority: :content_change,
+    title: :subscriber_list,
+    subscription_count: :itself,
+  }.freeze
+
+  def call(start_time: nil, end_time: nil)
+    start_time = start_time ? Time.zone.parse(start_time) : 1.week.ago
+    end_time = end_time ? Time.zone.parse(end_time) : Time.zone.now
+
+    subscription_count_query = Subscription
+      .select("count(*)")
+      .where("subscriptions.created_at <= content_changes.created_at")
+      .where("subscriptions.ended_at is null or subscriptions.ended_at > content_changes.created_at")
+      .where("subscriptions.subscriber_list_id = subscriber_lists.id")
+      .where(frequency: "immediately")
+      .arel
+      .as("subscription_count")
+
+    query = MatchedContentChange
+      .where("content_changes.created_at": start_time..end_time)
+      .includes(:subscriber_list, :content_change)
+      .order(subscription_count: :desc)
+      .select("content_changes.*", "subscriber_lists.*")
```

Comment:
Thanks for putting this in, I'm going to be a bit of a pain and suggest that we shouldn't actually have it in as Activerecord won't use any of the data (it's also missing MatchedContentChange too).

For example trying it out with and without the where we get:

with:

```
irb(main):004:0> MatchedContentChange.where("content_changes.created_at": (1.week.ago)..(Time.zone.now)).includes(:subscriber_list, :content_change).order(subscription_count: :desc).select("content_changes.*", "subscriber_lists.*").select(Subscription.select("count(*)").where("subscriptions.created_at <= content_changes.created_at").where("subscriptions.ended_at IS NULL OR subscriptions.ended_at > content_changes.created_at").where("subscriptions.subscriber_list_id = subscriber_lists.id").where(frequency: "immediately").arel.as("subscription_count"))
  SQL (19.8ms)  SELECT content_changes.*, subscriber_lists.*, (SELECT count(*) FROM "subscriptions" WHERE (subscriptions.created_at <= content_changes.created_at) AND (subscriptions.ended_at IS NULL OR subscriptions.ended_at > content_changes.created_at) AND (subscriptions.subscriber_list_id = subscriber_lists.id) AND "subscriptions"."frequency" = $1) subscription_count, "matched_content_changes"."id" AS t0_r0, "matched_content_changes"."subscriber_list_id" AS t0_r1, "matched_content_changes"."created_at" AS t0_r2, "matched_content_changes"."updated_at" AS t0_r3, "matched_content_changes"."content_change_id" AS t0_r4, "subscriber_lists"."id" AS t1_r0, "subscriber_lists"."title" AS t1_r1, "subscriber_lists"."created_at" AS t1_r2, "subscriber_lists"."updated_at" AS t1_r3, "subscriber_lists"."document_type" AS t1_r4, "subscriber_lists"."tags" AS t1_r5, "subscriber_lists"."links" AS t1_r6, "subscriber_lists"."email_document_supertype" AS t1_r7, "subscriber_lists"."government_document_supertype" AS t1_r8, "subscriber_lists"."signon_user_uid" AS t1_r9, "subscriber_lists"."slug" AS t1_r10, "subscriber_lists"."url" AS t1_r11, "subscriber_lists"."description" AS t1_r12, "subscriber_lists"."tags_digest" AS t1_r13, "subscriber_lists"."links_digest" AS t1_r14, "subscriber_lists"."group_id" AS t1_r15, "content_changes"."id" AS t2_r0, "content_changes"."content_id" AS t2_r1, "content_changes"."title" AS t2_r2, "content_changes"."base_path" AS t2_r3, "content_changes"."change_note" AS t2_r4, "content_changes"."description" AS t2_r5, "content_changes"."links" AS t2_r6, "content_changes"."tags" AS t2_r7, "content_changes"."public_updated_at" AS t2_r8, "content_changes"."email_document_supertype" AS t2_r9, "content_changes"."government_document_supertype" AS t2_r10, "content_changes"."govuk_request_id" AS t2_r11, "content_changes"."document_type" AS t2_r12, "content_changes"."publishing_app" AS t2_r13, "content_changes"."created_at" AS t2_r14, "content_changes"."updated_at" AS t2_r15, "content_changes"."processed_at" AS t2_r16, "content_changes"."priority" AS t2_r17, "content_changes"."signon_user_uid" AS t2_r18, "content_changes"."footnote" AS t2_r19 FROM "matched_content_changes" LEFT OUTER JOIN "subscriber_lists" ON "subscriber_lists"."id" = "matched_content_changes"."subscriber_list_id" LEFT OUTER JOIN "content_changes" ON "content_changes"."id" = "matched_content_changes"."content_change_id" WHERE "content_changes"."created_at" BETWEEN $2 AND $3 ORDER BY "subscription_count" DESC LIMIT $4  [["frequency", 0], ["created_at", "2020-06-04 10:39:40.843069"], ["created_at", "2020-06-11 10:39:40.848774"], ["LIMIT", 11]]
```

without:

```
  MatchedContentChange Load (0.7ms)  SELECT content_changes.*, subscriber_lists.*, (SELECT count(*) FROM "subscriptions" WHERE (subscriptions.created_at <= content_changes.created_at) AND (subscriptions.ended_at IS NULL OR subscriptions.ended_at > content_changes.created_at) AND (subscriptions.subscriber_list_id = subscriber_lists.id) AND "subscriptions"."frequency" = $1) subscription_count FROM "matched_content_changes" ORDER BY "subscription_count" DESC LIMIT $2  [["frequency", 0], ["LIMIT", 11]]
Traceback (most recent call last):
ActiveRecord::StatementInvalid (PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "content_changes")
LINE 1: SELECT content_changes.*, subscriber_lists.*, (SELECT count(...
```

Reply:
Oh dear, I was only looking at the SQL. It frustrates me that the behaviour was so fragile. I finally realised from your SQL comment that I'm wanting an outer JOIN.

I've updated with something that I hope is much more understandable. Would you like to take a look?

(I think I was conflating `includes` with `joins` and `left_joins`.)

Reply:
Great - the `select("*")` seems to have cleared up the fragility the where introduces 👍  I think however by losing the `includes` you're going to have an N+1 problem where ActiveRecord has to look up each of the associations.

If we're always going to apply a where on a content_change do we gain anything from a LEFT JOIN? 

Sorry, I feel like I've pulled you into a rather deep rabbit hole where there aren't many clear rewards.

Reply:
That's OK. I think it's worth it for the person who inevitably comes to read this in future. 

- That's a good point about the N+1. I've added a separate `includes` and two comments to explain the need for both.

- I also agree we can use a `joins` here, since every `MatchedContentChange` must have both of the things we're joining on.

I _think_ it's now perfect 😄.

Reply:
Amazing - thanks.